[Deepin-Kernel-SIG] [linux 6.6-y] [Deepin] deepin: arm64: cpufeature: disable LSE on some cpus

opsiff opened the pull request:
V2 of https://github.com/deepin-community/kernel/pull/1302

deepin inclusion
category: performance

Disable LSE (Large System Extension) atomic instructions on some systems to improve performance of per-CPU atomic operations. LSE atomics can exhibit significant overhead on certain microarchitectures (e.g., TSV110) due to "far atomic" implementations	bypassing L1 cache. LL/SC (Load-Link/Store-Conditional) is substantially faster.

The default value is 0 (enabled), which automatically disables LSE on some systems. Set to 1 to skip the check enablement on our test systems regardless of performance impact.

When this feature is active, the kernel logs:
"LSE atomics: use llsc for performance, use lse_disable_check=1 to disable the feature."

PS:
Test with byte-unixbench6 in kp920 24c and 64GB memory, improve whole scores by 3.8%.

Link: https://lore.kernel.org/r/e7d539ed-ced0-4b96-8ecd-048a5b803b85@paulmck-laptop [1]
Tested-by: liujiangtao <liujiangtao@uniontech.com>

## Summary by Sourcery

Gate use of ARM64 LSE atomic instructions behind a CPU-specific performance check and tunable kernel parameter.

New Features:
- Introduce a kernel parameter to control whether CPU-based disabling of LSE atomic instructions is applied at boot.

Enhancements:
- Disable LSE atomics in favor of LL/SC on specific arm64 CPUs with known LSE performance issues, with a boot-time log message indicating the behavior.

Documentation:
- Document the new kernel parameter for controlling LSE atomic instruction disabling behavior in the admin guide.

Avenger-285714 commented on arch/arm64/kernel/cpufeature.c:
```diff
@@ -1540,6 +1540,34 @@ static bool has_32bit_el0(const struct arm64_cpu_capabilities *entry, int scope)
+	/* List of CPUs that LSE are slow more than llsc*/
+	static const struct midr_range lse_disable_list[] = {
+		MIDR_ALL_VERSIONS(MIDR_HISI_TSV110),
+		MIDR_ALL_VERSIONS(MIDR_PHYTIUM_FTC862),
```
已知的还有FTC863和此芯？贵司内部应该还有apple m1、rk3588和全志的设备，以及yitian710的服务器测试机，建议也同步测试下？

opsiff commented:
D3000(FTC862)
uos@uos-PC:~/Downloads/debug/LSE$ ./percpu_bench 
ARM64 Per-CPU Atomic Add Benchmark
===================================
Running percentile measurements (100 iterations)...
Detected 8 CPUs

 CPU: 0 - Latency Percentiles:
====================
LSE (stadd)     (c                0, d                0):   p50: 036.07 ns        p95: 036.29 ns          p99: 036.29 ns
LSE (stadd)     (c                0, d              100):   p50: 042.47 ns        p95: 042.69 ns          p99: 042.70 ns
LSE (stadd)     (c                0, d              200):   p50: 082.75 ns        p95: 082.76 ns          p99: 082.78 ns
sched_setaffinity: Invalid argument
LSE (stadd)     (c               10, d                0):   p50: 072.29 ns        p95: 072.60 ns          p99: 072.70 ns
sched_setaffinity: Invalid argument
LSE (stadd)     (c               10, d              300):   p50: 245.85 ns        p95: 246.13 ns          p99: 246.34 ns
sched_setaffinity: Invalid argument
LSE (stadd)     (c               10, d              500):   p50: 416.96 ns        p95: 417.27 ns          p99: 417.32 ns
sched_setaffinity: Invalid argument
LSE (stadd)     (c               30, d                0):   p50: 072.29 ns        p95: 072.60 ns          p99: 072.78 ns
LL/SC           (c                0, d                0):   p50: 007.61 ns        p95: 007.83 ns          p99: 007.83 ns
LL/SC           (c                0, d               10):   p50: 009.22 ns        p95: 009.43 ns          p99: 009.44 ns
LL/SC           (c                0, d               20):   p50: 010.82 ns        p95: 011.04 ns          p99: 011.04 ns
sched_setaffinity: Invalid argument
LL/SC           (c               10, d                0):   p50: 015.12 ns        p95: 015.62 ns          p99: 015.63 ns
sched_setaffinity: Invalid argument
LL/SC           (c               10, d               10):   p50: 018.52 ns        p95: 018.73 ns          p99: 019.02 ns
sched_setaffinity: Invalid argument
LL/SC           (c               10, d               20):   p50: 021.62 ns        p95: 022.05 ns          p99: 022.12 ns
sched_setaffinity: Invalid argument
LL/SC           (c             1000, d                0):   p50: 015.12 ns        p95: 015.62 ns          p99: 015.82 ns
sched_setaffinity: Invalid argument
LL/SC           (c             1000, d               10):   p50: 018.52 ns        p95: 018.74 ns          p99: 018.94 ns
sched_setaffinity: Invalid argument
LL/SC           (c          1000000, d                0):   p50: 015.12 ns        p95: 015.62 ns          p99: 015.91 ns
sched_setaffinity: Invalid argument
LL/SC           (c          1000000, d               10):   p50: 018.52 ns        p95: 018.73 ns          p99: 018.74 ns
LDADD           (c                0, d                0):   p50: 005.61 ns        p95: 005.61 ns          p99: 005.82 ns
LDADD           (c                0, d              100):   p50: 042.47 ns        p95: 042.69 ns          p99: 042.69 ns
LDADD           (c                0, d              200):   p50: 082.75 ns        p95: 082.75 ns          p99: 082.77 ns
LDADD           (c                0, d              300):   p50: 122.81 ns        p95: 123.03 ns          p99: 123.03 ns
sched_setaffinity: Invalid argument
LDADD           (c                1, d               10):   p50: 014.42 ns        p95: 014.81 ns          p99: 014.91 ns
sched_setaffinity: Invalid argument
LDADD           (c               10, d                0):   p50: 011.31 ns        p95: 011.51 ns          p99: 011.71 ns
sched_setaffinity: Invalid argument
LDADD           (c               10, d               10):   p50: 014.42 ns        p95: 014.73 ns          p99: 014.93 ns
sched_setaffinity: Invalid argument
LDADD           (c              100, d                0):   p50: 011.31 ns        p95: 011.53 ns          p99: 011.71 ns
PFRM_KEEP+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.44 ns          p99: 005.58 ns
sched_setaffinity: Invalid argument
PFRM_KEEP+STADD (c               10, d                0):   p50: 010.41 ns        p95: 010.77 ns          p99: 010.90 ns
sched_setaffinity: Invalid argument
PFRM_KEEP+STADD (c             1000, d                0):   p50: 010.61 ns        p95: 011.38 ns          p99: 011.91 ns
sched_setaffinity: Invalid argument
PFRM_KEEP+STADD (c          1000000, d                0):   p50: 010.52 ns        p95: 010.85 ns          p99: 011.15 ns
PFRM_STRM+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.50 ns          p99: 005.58 ns
sched_setaffinity: Invalid argument
PFRM_STRM+STADD (c               10, d                0):   p50: 010.47 ns        p95: 010.83 ns          p99: 010.99 ns
sched_setaffinity: Invalid argument
PFRM_STRM+STADD (c             1000, d                0):   p50: 010.61 ns        p95: 010.85 ns          p99: 011.22 ns
sched_setaffinity: Invalid argument
PFRM_STRM+STADD (c          1000000, d                0):   p50: 010.41 ns        p95: 010.77 ns          p99: 010.95 ns

 CPU: 1 - Latency Percentiles:
====================
LSE (stadd)     (c                0, d                0):   p50: 036.06 ns        p95: 036.07 ns          p99: 036.08 ns
LSE (stadd)     (c                0, d              100):   p50: 042.49 ns        p95: 042.51 ns          p99: 042.52 ns
LSE (stadd)     (c                0, d              200):   p50: 082.53 ns        p95: 082.59 ns          p99: 082.59 ns
LSE (stadd)     (c               10, d                0):   p50: 065.93 ns        p95: 066.11 ns          p99: 066.12 ns
LSE (stadd)     (c               10, d              300):   p50: 122.58 ns        p95: 122.58 ns          p99: 122.60 ns
LSE (stadd)     (c               10, d              500):   p50: 207.90 ns        p95: 207.90 ns          p99: 207.93 ns
LSE (stadd)     (c               30, d                0):   p50: 065.93 ns        p95: 066.11 ns          p99: 066.11 ns
LL/SC           (c                0, d                0):   p50: 007.61 ns        p95: 007.62 ns          p99: 007.62 ns
LL/SC           (c                0, d               10):   p50: 009.21 ns        p95: 009.22 ns          p99: 009.22 ns
LL/SC           (c                0, d               20):   p50: 010.82 ns        p95: 010.82 ns          p99: 010.82 ns
LL/SC           (c               10, d                0):   p50: 043.04 ns        p95: 056.49 ns          p99: 063.72 ns
LL/SC           (c               10, d               10):   p50: 056.65 ns        p95: 071.03 ns          p99: 073.76 ns
LL/SC           (c               10, d               20):   p50: 035.75 ns        p95: 040.12 ns          p99: 043.52 ns
LL/SC           (c             1000, d                0):   p50: 017.37 ns        p95: 017.39 ns          p99: 017.40 ns
LL/SC           (c             1000, d               10):   p50: 023.31 ns        p95: 023.31 ns          p99: 023.31 ns
LL/SC           (c          1000000, d                0):   p50: 012.92 ns        p95: 013.59 ns          p99: 013.98 ns
LL/SC           (c          1000000, d               10):   p50: 016.80 ns        p95: 017.65 ns          p99: 017.92 ns
LDADD           (c                0, d                0):   p50: 005.61 ns        p95: 005.61 ns          p99: 005.61 ns
LDADD           (c                0, d              100):   p50: 042.47 ns        p95: 042.47 ns          p99: 042.48 ns
LDADD           (c                0, d              200):   p50: 082.53 ns        p95: 082.54 ns          p99: 082.55 ns
LDADD           (c                0, d              300):   p50: 122.60 ns        p95: 122.60 ns          p99: 122.62 ns
LDADD           (c                1, d               10):   p50: 050.63 ns        p95: 050.97 ns          p99: 050.98 ns
LDADD           (c               10, d                0):   p50: 050.28 ns        p95: 052.05 ns          p99: 052.05 ns
LDADD           (c               10, d               10):   p50: 049.51 ns        p95: 050.09 ns          p99: 050.12 ns
LDADD           (c              100, d                0):   p50: 033.71 ns        p95: 033.72 ns          p99: 033.73 ns
PFRM_KEEP+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.21 ns          p99: 005.29 ns
PFRM_KEEP+STADD (c               10, d                0):   p50: 056.11 ns        p95: 057.69 ns          p99: 057.97 ns
PFRM_KEEP+STADD (c             1000, d                0):   p50: 024.84 ns        p95: 025.10 ns          p99: 025.18 ns
PFRM_KEEP+STADD (c          1000000, d                0):   p50: 009.72 ns        p95: 009.92 ns          p99: 010.02 ns
PFRM_STRM+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.21 ns          p99: 005.24 ns
PFRM_STRM+STADD (c               10, d                0):   p50: 055.74 ns        p95: 057.62 ns          p99: 057.89 ns
PFRM_STRM+STADD (c             1000, d                0):   p50: 024.68 ns        p95: 025.01 ns          p99: 025.05 ns
PFRM_STRM+STADD (c          1000000, d                0):   p50: 009.84 ns        p95: 010.13 ns          p99: 010.27 ns

 CPU: 2 - Latency Percentiles:
====================
LSE (stadd)     (c                0, d                0):   p50: 035.60 ns        p95: 035.66 ns          p99: 035.75 ns
LSE (stadd)     (c                0, d              100):   p50: 042.47 ns        p95: 042.48 ns          p99: 042.49 ns
LSE (stadd)     (c                0, d              200):   p50: 082.54 ns        p95: 082.55 ns          p99: 082.55 ns
LSE (stadd)     (c               10, d                0):   p50: 065.74 ns        p95: 065.94 ns          p99: 065.95 ns
LSE (stadd)     (c               10, d              300):   p50: 122.58 ns        p95: 122.59 ns          p99: 122.62 ns
LSE (stadd)     (c               10, d              500):   p50: 207.91 ns        p95: 207.97 ns          p99: 207.99 ns
LSE (stadd)     (c               30, d                0):   p50: 065.73 ns        p95: 065.91 ns          p99: 065.92 ns
LL/SC           (c                0, d                0):   p50: 007.61 ns        p95: 007.62 ns          p99: 007.62 ns
LL/SC           (c                0, d               10):   p50: 009.22 ns        p95: 009.22 ns          p99: 009.22 ns
LL/SC           (c                0, d               20):   p50: 010.82 ns        p95: 010.82 ns          p99: 010.83 ns
LL/SC           (c               10, d                0):   p50: 037.72 ns        p95: 046.60 ns          p99: 051.77 ns
LL/SC           (c               10, d               10):   p50: 054.29 ns        p95: 069.08 ns          p99: 071.54 ns
LL/SC           (c               10, d               20):   p50: 051.40 ns        p95: 066.57 ns          p99: 069.01 ns
LL/SC           (c             1000, d                0):   p50: 017.52 ns        p95: 017.53 ns          p99: 017.55 ns
LL/SC           (c             1000, d               10):   p50: 023.28 ns        p95: 023.29 ns          p99: 023.30 ns
LL/SC           (c          1000000, d                0):   p50: 012.60 ns        p95: 013.65 ns          p99: 013.99 ns
LL/SC           (c          1000000, d               10):   p50: 016.52 ns        p95: 017.53 ns          p99: 018.06 ns
LDADD           (c                0, d                0):   p50: 005.61 ns        p95: 005.61 ns          p99: 005.61 ns
LDADD           (c                0, d              100):   p50: 042.47 ns        p95: 042.48 ns          p99: 042.49 ns
LDADD           (c                0, d              200):   p50: 082.54 ns        p95: 082.54 ns          p99: 082.55 ns
LDADD           (c                0, d              300):   p50: 122.60 ns        p95: 122.63 ns          p99: 122.68 ns
LDADD           (c                1, d               10):   p50: 050.63 ns        p95: 051.19 ns          p99: 051.20 ns
LDADD           (c               10, d                0):   p50: 050.28 ns        p95: 052.05 ns          p99: 052.05 ns
LDADD           (c               10, d               10):   p50: 050.08 ns        p95: 050.11 ns          p99: 050.12 ns
LDADD           (c              100, d                0):   p50: 033.67 ns        p95: 033.68 ns          p99: 033.70 ns
PFRM_KEEP+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.46 ns          p99: 005.77 ns
PFRM_KEEP+STADD (c               10, d                0):   p50: 053.03 ns        p95: 053.37 ns          p99: 053.55 ns
PFRM_KEEP+STADD (c             1000, d                0):   p50: 024.09 ns        p95: 024.38 ns          p99: 024.60 ns
PFRM_KEEP+STADD (c          1000000, d                0):   p50: 021.92 ns        p95: 023.14 ns          p99: 025.38 ns
PFRM_STRM+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.60 ns          p99: 006.07 ns
PFRM_STRM+STADD (c               10, d                0):   p50: 053.11 ns        p95: 053.47 ns          p99: 053.72 ns
PFRM_STRM+STADD (c             1000, d                0):   p50: 024.05 ns        p95: 024.71 ns          p99: 025.25 ns
PFRM_STRM+STADD (c          1000000, d                0):   p50: 022.23 ns        p95: 023.19 ns          p99: 024.64 ns

 CPU: 3 - Latency Percentiles:
====================
LSE (stadd)     (c                0, d                0):   p50: 036.07 ns        p95: 036.09 ns          p99: 036.18 ns
LSE (stadd)     (c                0, d              100):   p50: 042.47 ns        p95: 042.48 ns          p99: 042.49 ns
LSE (stadd)     (c                0, d              200):   p50: 082.54 ns        p95: 082.55 ns          p99: 082.56 ns
LSE (stadd)     (c               10, d                0):   p50: 065.96 ns        p95: 066.14 ns          p99: 066.15 ns
LSE (stadd)     (c               10, d              300):   p50: 122.59 ns        p95: 122.60 ns          p99: 122.61 ns
LSE (stadd)     (c               10, d              500):   p50: 207.92 ns        p95: 207.94 ns          p99: 207.97 ns
LSE (stadd)     (c               30, d                0):   p50: 065.95 ns        p95: 066.13 ns          p99: 066.14 ns
LL/SC           (c                0, d                0):   p50: 007.61 ns        p95: 007.62 ns          p99: 007.62 ns
LL/SC           (c                0, d               10):   p50: 009.22 ns        p95: 009.22 ns          p99: 009.23 ns
LL/SC           (c                0, d               20):   p50: 010.82 ns        p95: 010.83 ns          p99: 010.83 ns
LL/SC           (c               10, d                0):   p50: 042.39 ns        p95: 053.59 ns          p99: 055.96 ns
LL/SC           (c               10, d               10):   p50: 057.61 ns        p95: 072.14 ns          p99: 082.63 ns
LL/SC           (c               10, d               20):   p50: 041.44 ns        p95: 050.20 ns          p99: 053.50 ns
LL/SC           (c             1000, d                0):   p50: 017.44 ns        p95: 017.45 ns          p99: 017.47 ns
LL/SC           (c             1000, d               10):   p50: 023.32 ns        p95: 023.33 ns          p99: 023.33 ns
LL/SC           (c          1000000, d                0):   p50: 012.20 ns        p95: 013.27 ns          p99: 013.58 ns
LL/SC           (c          1000000, d               10):   p50: 016.56 ns        p95: 017.69 ns          p99: 018.03 ns
LDADD           (c                0, d                0):   p50: 005.61 ns        p95: 005.61 ns          p99: 005.62 ns
LDADD           (c                0, d              100):   p50: 042.47 ns        p95: 042.48 ns          p99: 042.49 ns
LDADD           (c                0, d              200):   p50: 082.54 ns        p95: 082.55 ns          p99: 082.56 ns
LDADD           (c                0, d              300):   p50: 122.61 ns        p95: 122.63 ns          p99: 122.65 ns
LDADD           (c                1, d               10):   p50: 050.45 ns        p95: 050.93 ns          p99: 050.94 ns
LDADD           (c               10, d                0):   p50: 050.85 ns        p95: 051.11 ns          p99: 051.12 ns
LDADD           (c               10, d               10):   p50: 050.04 ns        p95: 050.06 ns          p99: 050.08 ns
LDADD           (c              100, d                0):   p50: 033.70 ns        p95: 033.72 ns          p99: 033.73 ns
PFRM_KEEP+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.36 ns          p99: 005.38 ns
PFRM_KEEP+STADD (c               10, d                0):   p50: 055.25 ns        p95: 056.41 ns          p99: 056.51 ns
PFRM_KEEP+STADD (c             1000, d                0):   p50: 023.77 ns        p95: 024.05 ns          p99: 024.14 ns
PFRM_KEEP+STADD (c          1000000, d                0):   p50: 019.89 ns        p95: 021.64 ns          p99: 022.02 ns
PFRM_STRM+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.37 ns          p99: 005.45 ns
PFRM_STRM+STADD (c               10, d                0):   p50: 055.31 ns        p95: 056.33 ns          p99: 056.48 ns
PFRM_STRM+STADD (c             1000, d                0):   p50: 023.75 ns        p95: 024.09 ns          p99: 024.14 ns
PFRM_STRM+STADD (c          1000000, d                0):   p50: 020.31 ns        p95: 021.63 ns          p99: 022.04 ns

 CPU: 4 - Latency Percentiles:
====================
LSE (stadd)     (c                0, d                0):   p50: 038.74 ns        p95: 038.79 ns          p99: 038.90 ns
LSE (stadd)     (c                0, d              100):   p50: 042.47 ns        p95: 042.48 ns          p99: 042.49 ns
LSE (stadd)     (c                0, d              200):   p50: 082.54 ns        p95: 082.54 ns          p99: 082.55 ns
LSE (stadd)     (c               10, d                0):   p50: 051.08 ns        p95: 051.10 ns          p99: 051.16 ns
LSE (stadd)     (c               10, d              300):   p50: 122.59 ns        p95: 122.60 ns          p99: 122.62 ns
LSE (stadd)     (c               10, d              500):   p50: 207.93 ns        p95: 207.93 ns          p99: 207.94 ns
LSE (stadd)     (c               30, d                0):   p50: 051.07 ns        p95: 051.09 ns          p99: 051.10 ns
LL/SC           (c                0, d                0):   p50: 007.61 ns        p95: 007.62 ns          p99: 007.62 ns
LL/SC           (c                0, d               10):   p50: 009.21 ns        p95: 009.22 ns          p99: 009.22 ns
LL/SC           (c                0, d               20):   p50: 010.82 ns        p95: 010.82 ns          p99: 010.82 ns
LL/SC           (c               10, d                0):   p50: 068.31 ns        p95: 084.24 ns          p99: 087.64 ns
LL/SC           (c               10, d               10):   p50: 085.70 ns        p95: 103.82 ns          p99: 111.44 ns
LL/SC           (c               10, d               20):   p50: 112.97 ns        p95: 135.20 ns          p99: 139.94 ns
LL/SC           (c             1000, d                0):   p50: 020.65 ns        p95: 020.66 ns          p99: 020.66 ns
LL/SC           (c             1000, d               10):   p50: 024.78 ns        p95: 024.79 ns          p99: 024.79 ns
LL/SC           (c          1000000, d                0):   p50: 010.91 ns        p95: 011.71 ns          p99: 012.10 ns
LL/SC           (c          1000000, d               10):   p50: 012.95 ns        p95: 014.02 ns          p99: 014.36 ns
LDADD           (c                0, d                0):   p50: 005.61 ns        p95: 005.61 ns          p99: 005.61 ns
LDADD           (c                0, d              100):   p50: 042.47 ns        p95: 042.48 ns          p99: 042.48 ns
LDADD           (c                0, d              200):   p50: 082.54 ns        p95: 082.54 ns          p99: 082.55 ns
LDADD           (c                0, d              300):   p50: 122.60 ns        p95: 122.61 ns          p99: 122.63 ns
LDADD           (c                1, d               10):   p50: 054.64 ns        p95: 056.08 ns          p99: 056.10 ns
LDADD           (c               10, d                0):   p50: 052.39 ns        p95: 054.09 ns          p99: 054.11 ns
LDADD           (c               10, d               10):   p50: 054.64 ns        p95: 056.09 ns          p99: 056.15 ns
LDADD           (c              100, d                0):   p50: 067.59 ns        p95: 070.06 ns          p99: 070.06 ns
PFRM_KEEP+STADD (c                0, d                0):   p50: 006.38 ns        p95: 007.02 ns          p99: 007.49 ns
PFRM_KEEP+STADD (c               10, d                0):   p50: 052.44 ns        p95: 052.69 ns          p99: 052.75 ns
PFRM_KEEP+STADD (c             1000, d                0):   p50: 039.45 ns        p95: 039.60 ns          p99: 039.65 ns
PFRM_KEEP+STADD (c          1000000, d                0):   p50: 035.43 ns        p95: 036.70 ns          p99: 037.21 ns
PFRM_STRM+STADD (c                0, d                0):   p50: 006.34 ns        p95: 007.18 ns          p99: 007.82 ns
PFRM_STRM+STADD (c               10, d                0):   p50: 052.36 ns        p95: 052.73 ns          p99: 052.80 ns
PFRM_STRM+STADD (c             1000, d                0):   p50: 039.40 ns        p95: 039.60 ns          p99: 039.63 ns
PFRM_STRM+STADD (c          1000000, d                0):   p50: 035.10 ns        p95: 036.80 ns          p99: 036.96 ns

 CPU: 5 - Latency Percentiles:
====================
LSE (stadd)     (c                0, d                0):   p50: 038.74 ns        p95: 038.76 ns          p99: 038.87 ns
LSE (stadd)     (c                0, d              100):   p50: 042.47 ns        p95: 042.47 ns          p99: 042.49 ns
LSE (stadd)     (c                0, d              200):   p50: 082.54 ns        p95: 082.54 ns          p99: 082.56 ns
LSE (stadd)     (c               10, d                0):   p50: 051.05 ns        p95: 051.10 ns          p99: 051.14 ns
LSE (stadd)     (c               10, d              300):   p50: 122.59 ns        p95: 122.60 ns          p99: 122.62 ns
LSE (stadd)     (c               10, d              500):   p50: 207.92 ns        p95: 207.92 ns          p99: 207.93 ns
LSE (stadd)     (c               30, d                0):   p50: 051.08 ns        p95: 051.10 ns          p99: 051.11 ns
LL/SC           (c                0, d                0):   p50: 007.61 ns        p95: 007.62 ns          p99: 007.62 ns
LL/SC           (c                0, d               10):   p50: 009.21 ns        p95: 009.22 ns          p99: 009.22 ns
LL/SC           (c                0, d               20):   p50: 010.82 ns        p95: 010.82 ns          p99: 010.82 ns
LL/SC           (c               10, d                0):   p50: 077.33 ns        p95: 102.00 ns          p99: 108.84 ns
LL/SC           (c               10, d               10):   p50: 039.73 ns        p95: 048.25 ns          p99: 050.73 ns
LL/SC           (c               10, d               20):   p50: 050.92 ns        p95: 061.09 ns          p99: 064.16 ns
LL/SC           (c             1000, d                0):   p50: 020.68 ns        p95: 020.68 ns          p99: 020.69 ns
LL/SC           (c             1000, d               10):   p50: 024.81 ns        p95: 024.82 ns          p99: 024.82 ns
LL/SC           (c          1000000, d                0):   p50: 013.21 ns        p95: 014.63 ns          p99: 015.16 ns
LL/SC           (c          1000000, d               10):   p50: 016.16 ns        p95: 018.34 ns          p99: 018.89 ns
LDADD           (c                0, d                0):   p50: 005.61 ns        p95: 005.61 ns          p99: 005.61 ns
LDADD           (c                0, d              100):   p50: 042.47 ns        p95: 042.48 ns          p99: 042.49 ns
LDADD           (c                0, d              200):   p50: 082.54 ns        p95: 082.54 ns          p99: 082.55 ns
LDADD           (c                0, d              300):   p50: 122.60 ns        p95: 122.61 ns          p99: 122.63 ns
LDADD           (c                1, d               10):   p50: 054.59 ns        p95: 056.04 ns          p99: 056.05 ns
LDADD           (c               10, d                0):   p50: 052.64 ns        p95: 054.08 ns          p99: 054.15 ns
LDADD           (c               10, d               10):   p50: 054.58 ns        p95: 056.03 ns          p99: 056.10 ns
LDADD           (c              100, d                0):   p50: 067.59 ns        p95: 070.06 ns          p99: 070.07 ns
PFRM_KEEP+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.21 ns          p99: 005.21 ns
PFRM_KEEP+STADD (c               10, d                0):   p50: 050.98 ns        p95: 052.78 ns          p99: 052.91 ns
PFRM_KEEP+STADD (c             1000, d                0):   p50: 036.49 ns        p95: 036.88 ns          p99: 036.95 ns
PFRM_KEEP+STADD (c          1000000, d                0):   p50: 010.84 ns        p95: 011.12 ns          p99: 011.20 ns
PFRM_STRM+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.21 ns          p99: 005.21 ns
PFRM_STRM+STADD (c               10, d                0):   p50: 052.48 ns        p95: 052.84 ns          p99: 052.88 ns
PFRM_STRM+STADD (c             1000, d                0):   p50: 036.25 ns        p95: 036.66 ns          p99: 036.74 ns
PFRM_STRM+STADD (c          1000000, d                0):   p50: 010.86 ns        p95: 011.01 ns          p99: 011.30 ns

 CPU: 6 - Latency Percentiles:
====================
LSE (stadd)     (c                0, d                0):   p50: 038.73 ns        p95: 038.75 ns          p99: 038.89 ns
LSE (stadd)     (c                0, d              100):   p50: 042.47 ns        p95: 042.47 ns          p99: 042.48 ns
LSE (stadd)     (c                0, d              200):   p50: 082.54 ns        p95: 082.54 ns          p99: 082.55 ns
LSE (stadd)     (c               10, d                0):   p50: 051.06 ns        p95: 051.10 ns          p99: 051.11 ns
LSE (stadd)     (c               10, d              300):   p50: 122.59 ns        p95: 122.59 ns          p99: 122.61 ns
LSE (stadd)     (c               10, d              500):   p50: 207.92 ns        p95: 207.92 ns          p99: 207.96 ns
LSE (stadd)     (c               30, d                0):   p50: 051.07 ns        p95: 051.09 ns          p99: 051.15 ns
LL/SC           (c                0, d                0):   p50: 007.61 ns        p95: 007.62 ns          p99: 007.62 ns
LL/SC           (c                0, d               10):   p50: 009.21 ns        p95: 009.22 ns          p99: 009.22 ns
LL/SC           (c                0, d               20):   p50: 010.82 ns        p95: 010.82 ns          p99: 010.82 ns
LL/SC           (c               10, d                0):   p50: 083.18 ns        p95: 110.82 ns          p99: 117.84 ns
LL/SC           (c               10, d               10):   p50: 100.42 ns        p95: 126.74 ns          p99: 145.15 ns
LL/SC           (c               10, d               20):   p50: 048.80 ns        p95: 064.13 ns          p99: 067.95 ns
LL/SC           (c             1000, d                0):   p50: 020.65 ns        p95: 020.66 ns          p99: 020.66 ns
LL/SC           (c             1000, d               10):   p50: 024.78 ns        p95: 024.79 ns          p99: 024.79 ns
LL/SC           (c          1000000, d                0):   p50: 013.38 ns        p95: 014.58 ns          p99: 014.86 ns
LL/SC           (c          1000000, d               10):   p50: 016.19 ns        p95: 017.47 ns          p99: 018.07 ns
LDADD           (c                0, d                0):   p50: 005.61 ns        p95: 005.61 ns          p99: 005.61 ns
LDADD           (c                0, d              100):   p50: 042.47 ns        p95: 042.47 ns          p99: 042.48 ns
LDADD           (c                0, d              200):   p50: 082.53 ns        p95: 082.54 ns          p99: 082.55 ns
LDADD           (c                0, d              300):   p50: 122.60 ns        p95: 122.60 ns          p99: 122.63 ns
LDADD           (c                1, d               10):   p50: 054.25 ns        p95: 055.68 ns          p99: 055.68 ns
LDADD           (c               10, d                0):   p50: 052.23 ns        p95: 054.07 ns          p99: 054.11 ns
LDADD           (c               10, d               10):   p50: 054.43 ns        p95: 055.67 ns          p99: 055.68 ns
LDADD           (c              100, d                0):   p50: 067.59 ns        p95: 070.05 ns          p99: 070.06 ns
PFRM_KEEP+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.21 ns          p99: 007.22 ns
PFRM_KEEP+STADD (c               10, d                0):   p50: 051.37 ns        p95: 051.56 ns          p99: 051.58 ns
PFRM_KEEP+STADD (c             1000, d                0):   p50: 039.47 ns        p95: 039.58 ns          p99: 039.60 ns
PFRM_KEEP+STADD (c          1000000, d                0):   p50: 035.77 ns        p95: 037.46 ns          p99: 037.68 ns
PFRM_STRM+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.21 ns          p99: 008.18 ns
PFRM_STRM+STADD (c               10, d                0):   p50: 051.35 ns        p95: 051.58 ns          p99: 051.69 ns
PFRM_STRM+STADD (c             1000, d                0):   p50: 039.24 ns        p95: 039.44 ns          p99: 039.48 ns
PFRM_STRM+STADD (c          1000000, d                0):   p50: 035.80 ns        p95: 037.50 ns          p99: 037.73 ns

 CPU: 7 - Latency Percentiles:
====================
LSE (stadd)     (c                0, d                0):   p50: 038.74 ns        p95: 038.77 ns          p99: 038.83 ns
LSE (stadd)     (c                0, d              100):   p50: 042.47 ns        p95: 042.48 ns          p99: 042.49 ns
LSE (stadd)     (c                0, d              200):   p50: 082.54 ns        p95: 082.55 ns          p99: 082.56 ns
LSE (stadd)     (c               10, d                0):   p50: 051.08 ns        p95: 051.10 ns          p99: 051.18 ns
LSE (stadd)     (c               10, d              300):   p50: 122.64 ns        p95: 122.66 ns          p99: 122.66 ns
LSE (stadd)     (c               10, d              500):   p50: 207.93 ns        p95: 207.94 ns          p99: 207.96 ns
LSE (stadd)     (c               30, d                0):   p50: 051.07 ns        p95: 051.10 ns          p99: 051.19 ns
LL/SC           (c                0, d                0):   p50: 007.61 ns        p95: 007.62 ns          p99: 007.62 ns
LL/SC           (c                0, d               10):   p50: 009.22 ns        p95: 009.22 ns          p99: 009.22 ns
LL/SC           (c                0, d               20):   p50: 010.82 ns        p95: 010.83 ns          p99: 010.83 ns
LL/SC           (c               10, d                0):   p50: 073.82 ns        p95: 088.67 ns          p99: 097.72 ns
LL/SC           (c               10, d               10):   p50: 090.54 ns        p95: 114.39 ns          p99: 120.27 ns
LL/SC           (c               10, d               20):   p50: 050.41 ns        p95: 061.19 ns          p99: 064.25 ns
LL/SC           (c             1000, d                0):   p50: 020.66 ns        p95: 020.67 ns          p99: 020.67 ns
LL/SC           (c             1000, d               10):   p50: 024.79 ns        p95: 024.81 ns          p99: 024.81 ns
LL/SC           (c          1000000, d                0):   p50: 013.09 ns        p95: 014.74 ns          p99: 015.20 ns
LL/SC           (c          1000000, d               10):   p50: 015.31 ns        p95: 016.96 ns          p99: 017.78 ns
LDADD           (c                0, d                0):   p50: 005.61 ns        p95: 005.62 ns          p99: 005.62 ns
LDADD           (c                0, d              100):   p50: 042.47 ns        p95: 042.48 ns          p99: 042.49 ns
LDADD           (c                0, d              200):   p50: 082.54 ns        p95: 082.54 ns          p99: 082.56 ns
LDADD           (c                0, d              300):   p50: 122.61 ns        p95: 122.62 ns          p99: 122.63 ns
LDADD           (c                1, d               10):   p50: 054.65 ns        p95: 056.09 ns          p99: 056.14 ns
LDADD           (c               10, d                0):   p50: 052.35 ns        p95: 054.08 ns          p99: 054.12 ns
LDADD           (c               10, d               10):   p50: 054.62 ns        p95: 056.06 ns          p99: 056.07 ns
LDADD           (c              100, d                0):   p50: 067.53 ns        p95: 070.02 ns          p99: 070.04 ns
PFRM_KEEP+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.21 ns          p99: 005.22 ns
PFRM_KEEP+STADD (c               10, d                0):   p50: 051.74 ns        p95: 052.82 ns          p99: 052.89 ns
PFRM_KEEP+STADD (c             1000, d                0):   p50: 036.77 ns        p95: 037.16 ns          p99: 037.36 ns
PFRM_KEEP+STADD (c          1000000, d                0):   p50: 011.68 ns        p95: 013.88 ns          p99: 014.45 ns
PFRM_STRM+STADD (c                0, d                0):   p50: 005.21 ns        p95: 005.22 ns          p99: 005.29 ns
PFRM_STRM+STADD (c               10, d                0):   p50: 051.74 ns        p95: 052.77 ns          p99: 052.82 ns
PFRM_STRM+STADD (c             1000, d                0):   p50: 036.43 ns        p95: 037.04 ns          p99: 037.35 ns
PFRM_STRM+STADD (c          1000000, d                0):   p50: 011.86 ns        p95: 013.74 ns          p99: 014.00 ns

=== Benchmark Complete ===

opsiff committed:
deepin: arm64: cpufeature: disable LSE on some cpus

deepin inclusion
category: performance

Disable LSE (Large System Extensions) atomic instructions
on some systems to improve performance of per-CPU atomic
operations. LSE atomics can exhibit significant overhead
on certain microarchitectures (e.g., TSV110) due
to "far atomic" implementations	bypassing L1 cache. LL/SC
(Load-Link/Store-Conditional) is substantially faster.

The default value is 0 (enabled), which automatically
disables LSE on some systems. Set to 1 to skip the check
enablement on our test systems regardless of performance
impact.

When this feature is active, the kernel logs:
"LSE atomics: use llsc for performance, use lse_disable_check=1 to disable the feature."

PS:
Test with byte-unixbench6 in kp920 24c and 64GB memory,
improve whole scores by 3.8%.

Link: https://lore.kernel.org/r/e7d539ed-ced0-4b96-8ecd-048a5b803b85@paulmck-laptop [1]
Signed-off-by: Wentao Guan <guanwentao@uniontech.com>

opsiff commented on arch/arm64/kernel/cpufeature.c:
```diff
@@ -1540,6 +1540,34 @@ static bool has_32bit_el0(const struct arm64_cpu_capabilities *entry, int scope)
+
+static int __init arm64_lse_disable_check(char *str)
+{
+	return kstrtobool(str, &lse_disable_check);
```
这个是其他地方都这么使用的惯例 没必要增加代码
